crates/client/sync/src/commitments/lib.rs

```diff
@@ -146,109 +162,130 @@ where
-            let csd_clone = Arc::clone(&csd);
-            let bonsai_dbs_clone = Arc::clone(&bonsai_dbs);
-
-            task_set.spawn(contract_trie_root_loop(csd_clone, bonsai_dbs_clone, *contract_address, *class_hash));
-        });
-
-        task_set
-    })
-    .await?;
-
-    // The order in which contract trie roots are waited for is not important since each call to
-    // `update_contract_trie` in `contract_trie_root` mutates the Deoxys db.
-    let mut contract_trie_root = Felt252Wrapper::ZERO;
-    while let Some(res) = task_set.join_next().await {
-        contract_trie_root = match res? {
-            Ok(trie_root) => trie_root,
-            Err(e) => {
-                task_set.abort_all();
-                return Err(e);
-            }
-        }
+/// Calculates the contract trie root
+///
+/// # Arguments
+///
+/// * `csd`             - Commitment state diff for the current block.
+/// * `overrides`       - Deoxys storage override for accessing the Substrate db.
+/// * `bonsai_contract` - Bonsai db used to store contract hashes.
+/// * `block_number`    - The current block number.
+///
+/// # Returns
+///
+/// The contract root.
+fn contract_trie_root<B: BlockT>(
+    csd: &CommitmentStateDiff,
+    overrides: Arc<OverrideHandle<Block<Header<u32, BlakeTwo256>, OpaqueExtrinsic>>>,
+    mut bonsai_contract: MutexGuard<BonsaiStorage<BasicId, BonsaiDb<B>, Pedersen>>,
+    block_number: u64,
+    maybe_block_hash: Option<H256>,
+) -> Result<Felt252Wrapper, BonsaiStorageError<BonsaiDbError>> {
+    for (contract_address, updates) in csd.storage_updates.iter() {
+        let class_commitment_leaf_hash =
+            contract_state_leaf_hash(csd, overrides.as_ref(), contract_address, updates, maybe_block_hash)?;
+
+        let key = key(contract_address.0.0);
+        bonsai_contract.insert(&key, &class_commitment_leaf_hash.into())?;
     }
 
-    Ok(contract_trie_root)
+    bonsai_contract.commit(BasicId::new(block_number))?;
+    Ok(bonsai_contract.root_hash()?.into())
 }
 
-async fn contract_trie_root_loop<B: BlockT>(
-    csd: Arc<CommitmentStateDiff>,
-    bonsai_dbs: Arc<BonsaiDbs<B>>,
-    contract_address: ContractAddress,
-    class_hash: ClassHash,
-) -> anyhow::Result<Felt252Wrapper> {
-    let storage_root =
-        update_storage_trie(&contract_address, &csd, &bonsai_dbs.storage).expect("Failed to update storage trie");
-    let nonce = *csd.address_to_nonce.get(&contract_address).unwrap_or(&Felt252Wrapper::default().into());
+fn contract_state_leaf_hash(
+    csd: &CommitmentStateDiff,
+    overrides: &OverrideHandle<Block<Header<u32, BlakeTwo256>, OpaqueExtrinsic>>,
+    contract_address: &ContractAddress,
+    storage_updates: &IndexMap<StorageKey, StarkFelt>,
+    maybe_block_hash: Option<H256>,
+) -> Result<Felt252Wrapper, BonsaiStorageError<BonsaiDbError>> {
+    let storage_root = update_storage_trie(overrides, contract_address, storage_updates, maybe_block_hash)?;
+
+    let nonce =
```

Comment:
Is an error transformed into default?

Reply:
What do you mean by that?

If you are wondering whether conversion is performed automatically on errors caught with `?`, it is.

Reply:
it's about `.unwrap_or(..)`

Reply:
it's about `.unwrap_or(..)`

Reply:
That's some old code that was refactored. Waiting for v13 compatibility to be able to test removing it.